Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -196,4 +197,139 @@ lemma two_mul_semilength_eq_length : 2 * p.semilength = p.toList.length := by
 
 end Semilength
 
+section FirstReturn
+
+/-- `p.firstReturn` is 0 if `p = 0` and the index of the `D` matching the initial `U` otherwise. -/
+def firstReturn : ℕ :=
+  (range p.toList.length).findIdx (fun i ↦
+    (p.toList.take (i + 1)).count U = (p.toList.take (i + 1)).count D)
+
+@[simp] lemma firstReturn_zero : firstReturn 0 = 0 := rfl
+
+lemma firstReturn_pos (h : p.toList ≠ []) : 0 < p.firstReturn := by
+  by_contra f
+  replace f : p.firstReturn = 0 := by omega
+  rw [firstReturn, findIdx_eq (by rw [length_range]; exact length_pos.mpr h)] at f
+  simp only [get_eq_getElem, getElem_range] at f
+  rw [← p.cons_tail_dropLast_concat h] at f
+  simp at f
+
+lemma firstReturn_lt_length (h : p.toList ≠ []) : p.firstReturn < p.toList.length := by
+  have lp := length_pos_of_ne_nil h
+  rw [← length_range p.toList.length]
+  apply findIdx_lt_length_of_exists
+  simp only [mem_range, decide_eq_true_eq]
+  use p.toList.length - 1
+  exact ⟨by omega, by rw [Nat.sub_add_cancel lp, take_of_length_le (le_refl _),
+    p.count_U_eq_count_D]⟩
+
+lemma count_eq_count_of_take_firstReturn_add_one (h : p.toList ≠ []) :
```

Comment:
`take_firstReturn_add_one` is not an assumption, so this is a bad lemma name. What about this instead?
```suggestion
lemma count_take_firstReturn_add_one (h : p.toList ≠ []) :
```
The full name would be something like
```suggestion
lemma count_take_firstReturn_add_one_U_eq_count_take_firstReturn_add_one_D (h : p.toList ≠ []) :
```
but that's too long to be memorable